### ISSUE TITLE
XSPEC module is specific to a single patch / should autogenerate the interface code

### ISSUE THREAD

**Issue by DougBurke** [sherpa/sherpa#52]
I believe that a lot of the XSPEC module code - i.e. most of `_xspec.cc` and `__init__.py` - should be auto-generated based on the contents of the `model.dat` file in the XSPEC distribution. The main advantage to this approach is that it makes it possible for users to compile against the XSPEC module that they are using. That is, handle changes to the `model.dat` file that includes:

- parameter changes (e.g. a change to the min/max range)
- changes to the C/C++/Fortran model function being called (e.g. if the model has been re-written and
      now uses a different function)
- new models
- a more-consistent naming scheme for parameter names that don't map to Python (there is no
  existing scheme, as far as I can tell, and it looks like different rules were used for different models)
- removal of old models (not sure this has happened, but it could)

At present, users are forced to use an interface that assumes a single patch level - at present it's 12.8.2e - which should work for any 12.8.2 build, but leaves the user in a slightly-strange place if used against a different patch level (i.e. what bugs have/have not been fixed). If there's a 12.8.3 with new models then they will not pick them up until the code has been manually edited (which we have found to be problematic), and then there may be problems compiling the module against an older version of XSPEC.
 
There are some down sides to auto-generating most of the code - such as documentation, complicating the build process, and tracking down bugs (although I'm less bothered about the last point since the code that needs to be generated is quite simple).

I'd also like to expand access to the "utility" functions in XSPEC (in particular the keyword-related ones that *might* allow us to support some of the "mixing" models), which is related (some of them can be auto-generated, if there's an easy way to check for the functionality being present), but isn't an essential part of this request.

I plan to work on this, and have existing code that I can use as a starting point, but that's going to take a bit of time to put together.

**Comment by DougBurke** [comment]
Ahhh, there's a new release of XSPEC being tested, so I guess this is more relevant that I realised this morning!

**Comment by anetasie** [comment]
Yes. The new patch was just released and include the updates to [NEI models](https://heasarc.gsfc.nasa.gov/xanadu/xspec/issues/issues.html)

**Comment by DougBurke** [comment]
Aneta, I was talking about the next release, which is in the testing stage but not-yet released, not the latest patch.

**Comment by DougBurke** [comment]
Note that Xspec 12.9.0 has now been released. It contains one new model, a few changes to the API functions (none relevant to the functions we currently use), and changes in the parameter names (e.g. see #74) and values.

**Comment by DougBurke** [comment]
I have started a branch over at https://github.com/DougBurke/sherpa/tree/feature/build-xspec-module to explore auto-generating the interface to the model.dat models. It re-uses code from the code I wrote to support Xspec user models in Sherpa, rather than the existing code in the `scripts/` directory, as I was much-more familiar with the former.

At present the branch just permits the creation of code; it does not try to include the code in the build, as there needs to be some manual verification that the code does not diverge from the hand-edited code currently in `sherpa/astro/xspec/__init__.py` and `sherpa/astro/xspec/src/_xspec.cc`.

**Comment by DougBurke** [comment]
So, with recent changes we can see that the `_xspec.cc` file - at least the section between `// Start model definitions` and ``// End model definitions`` - can be autogenerated, which would allow the code to build on new versions that change the name of the function used by a model (such as the case for `rgsxsrc` with XSPEC 12.15.0 / #2252).

However, the code can likely not be imported because the `__init__.py` file will try to reference the function directly, **unless** we go and change all models from setting `_calc =  _xspec.<name>` to `__function__ = "<name>"`. Is there a run-time cost to this change or is it converted during model initialization (hopefully the latter)?

**Comment by DougBurke** [comment]
I tried in HEACIT to push people to using toml (or equivalent) for better marking up X-ray models and there was essentially no interest.

I have wanted this for a long time - this ticket was written almost 10 years ago! - but I just do not see it happening without a loss of documentation. 

We need two parts

- sherpa/astro/xspec/src/_xspec.cc
- sherpa/astro/xspec/__init__.py

We can autogenerate parts of `_xspec.cc`, but we can not auto generate all of it. The hope is the manually-created part will remain working, which has been generally fine but I'm pretty sure that has been at least one time it hasn't.

We can autogenerate parts of `__init__.py` but only with a huge loss in how useful the documentation is. I am not ready to give up on that yet.

There is also the fact that my previous experiences in auto-generating Python code and having it work well with setuptools has been less than spectacular (e.g. in the "I've tweaked the generator code, and I hope that the build recognizes this and rebuilds everything but doesn't so I end up wasting time because I am not actually testing the code I think I'm testing" type of situation). My hope is that the tracking done by `meson` would help here, but I am not willing to put any money on that, and also it would require changing the build backend, which has its own issues.

Another idea I have pursued is making the XSPEC interface - at least, the compiled code - be provided by an external module, and then Sherpa can just deal with the Python model classes - e.g. https://github.com/cxcsds/xspec-models-cxc - but again there has been essentially no bye-in from interested parties (other than "can I use this yet") and it still doesn't help with the Python side.
